### PR TITLE
refactor the scripts for rendering samples

### PR DESCRIPTION
**Pull request by reSHARMA**
Hello,

Sorry for the delayed response I was traveling, but I found  time to rethink about the whole code.
This pr is still under development, and test.sh is my copy of render.sh which I run on my system to debug.

At the end test.sh will get merged to render.sh

This would solve both the open issues :)

Thanks

### CODE REVIEW

**Comment by reSHARMA** [comment]
So, the default-version is gone. Say we want to freeze version x for dir abc then just add a file abc.txt in it :)

File: render.sh [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+
+get-synfig-appimage () {
+	SYNFIG="/tmp/synfig-$VERSION.appimage --appimage-exec synfig" 
+	if [ ! -d "/tmp/synfig-$VERSION.appimage" ]; then
+		if [ "$#" -eq 3 ]; then
+			LINK=$3
+		else
+			LINK="https://sourceforge.net/projects/synfig/files/releases/$VERSION/linux/synfigstudio-$2.appimage/download"
+		fi
+		wget --quiet $LINK -O "/tmp/synfig-$VERSION.appimage"
```
Comment (morevnaproject) [inline render.sh]: Please make sure to include changes from this commit - 37a89a05cf68d9c710ccea44c809742b489e9bdf

File: render.sh [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+	if [ $PARSED_VERSION -lt 120 ]; then
+		get-synfig-tar $VERSION
+	elif [ $PARSED_VERSION -eq 120 ]; then
+		get-synfig-appimage $VERSION "1.2.0-64bit-r2"
+	elif [ $PARSED_VERSION -eq 121 ]; then
+		get-synfig-appimage $VERSION "1.2.1-64bit"
+	elif [ $PARSED_VERSION -eq 122 ]; then
+		get-synfig-appimage $VERSION "18.09.14-linux64-286f1" "https://sourceforge.net/projects/synfig/files/releases/1.2.2/SynfigStudio-1.2.2-18.09.14-linux64-286f1.appimage/download"
+	elif [ $PARSED_VERSION -eq 131020181026 ]; then
```
Comment (morevnaproject) [inline render.sh]: Please make sure to update version as here  ^__^ - 852146c077e6931b18235aec7d83c52b32834faf

File: render.sh [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+	if [ ! -d "/tmp/synfig-$VERSION" ]; then
+		wget --quiet "https://sourceforge.net/projects/synfig/files/releases/$VERSION/linux/synfigstudio-$VERSION.x86_64.tar.bz2/download" -O "/tmp/synfig-$VERSION.tar.bz2"
+		mkdir -p /tmp/synfig-$VERSION
+		tar jxf /tmp/synfig-$VERSION.tar.bz2 -C /tmp/synfig-$VERSION --strip-components=1
+	fi
+}
+
+get-synfig-appimage () {
+	SYNFIG="/tmp/synfig-$VERSION.appimage --appimage-exec synfig" 
+	if [ ! -d "/tmp/synfig-$VERSION.appimage" ]; then
```
Comment (morevnaproject) [inline render.sh]: A quick fix - 1e223b9f8cd6e5e6bcbe4cc85ce8f18a609609e9

File: render.sh [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+	DIR=${file%/*}
+	NAME=${file##*/}
+	NAME=${NAME%.*}
+
+	# eg: sources/layers/rectangle/rectangle-0.sif
+	# EXT = "sif"
+	# DIR = "sources/layers/rectangle"
+	# NAME = "rectangle-0"
+
+	if [ $(echo $NAME | grep -i '-') ]; then
```
Comment (morevnaproject) [inline render.sh]: Clever! But that disallows to use "-" in directory names. ^_^ Someone (like me) can forget about that restriction someday. ^__^ So I am proposing something like this (not tested):

```
if [ "$EXT" = "sif" ]; then
   render-only-one-file $file
else [ "$EXT" = "txt"]; then
   PARENT_DIR_NAME=${DIR##*/}
   if [ "$PARENT_DIR_NAME" = "$NAME" ]; then
      render-only-one-dir $file
   else
      render-only-one-dir $file
   fi
fi
```

File: render.sh [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+render-only-one-dir () {
+	FILE=$1
+	DIR=${FILE%/*}
+	render-dir $DIR
+}
+
+# Get the modified files from the commit 
+
+CHANGED_FILES=`git diff --name-only $TRAVIS_COMMIT_RANGE`
```
Comment (morevnaproject) [inline render.sh]: I suggest to  add possibility to run the script locally, outside of Travis. All we need to do is to replace this line with the code like this (not tested):

```
if [ -z "$TRAVIS_BUILD_DIR" ]; then
   TRAVIS_BUILD_DIR=$(cd `dirname "$0"`; pwd) # gets directory of current script
fi
cd "$TRAVIS_BUILD_DIR"
if [ -z "$TRAVIS_COMMIT_RANGE" ]; then
   # rebuild everything
   CHANGED_FILES="sources/sources.txt"
else
   CHANGED_FILES=`git diff --name-only $TRAVIS_COMMIT_RANGE`
fi
```

**Comment by morevnaproject** [comment]
Great, thank you! I have made a quick review - please check. ^__^

**Comment by reSHARMA** [comment]
Thanks :)

Another issue would be that circle.txt would be the version for whole circle dir and also for circle.sif, so it is better to change the default name of sample to circle-default.sif or anything that you may suggest.

**Review by morevnaproject — CHANGES_REQUESTED** [review]

**Comment by morevnaproject** [comment]
> I guess i'm missing something, where should I look for review :)

Oh, sorry, I forgot to press "Submit review" button. Should be here (above) now.

> Another issue would be that circle.txt would be the version for whole circle dir and also for circle.sif, so it is better to change the default name of sample to circle-default.sif or anything that you may suggest.

Good catch! Fixed now - dd79af7a2b9097d783cb6eb827a14450f32ecfdb

File: render.sh (hunk 4 above)
Comment (reSHARMA) [reply render.sh]: Thanks for the review, I've added everything else and it is looking good and will add this one after testing :)

Comment (reSHARMA) [reply render.sh]: Awesome observation :)

**Comment by reSHARMA** [comment]
I think we are good to go :)

**Comment by morevnaproject** [comment]
Awesome, thank you! I have merged the changes and It seems to successfully parsed by Travis! - https://travis-ci.org/synfig/synfig-tests-regressions/builds/450962444 ^__^

I think we can consider this task done. ^__^

**Pull request merged**